flink/v1.15/flink/src/test/java/org/apache/iceberg/flink/sink/TestPartitionUdf.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.sink;
+
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.util.List;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.table.api.TableEnvironment;
+import org.apache.flink.table.api.TableResult;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.CloseableIterator;
+import org.apache.iceberg.expressions.Literal;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.transforms.Transform;
+import org.apache.iceberg.transforms.Transforms;
+import org.apache.iceberg.types.Types;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class TestPartitionUdf {
+
+  public static TableEnvironment tEnv;
+
+  @BeforeClass
+  public static void before() {
+    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    tEnv = StreamTableEnvironment.create(env);
+    tEnv.createTemporarySystemFunction("buckets", PartitionTransformUdf.Bucket.class);
+    tEnv.createTemporarySystemFunction("truncates", PartitionTransformUdf.Truncate.class);
```

Comment:
Question: Can we provide a single registration option so things are uniform? The `bucket` and `truncate` functions are universally needed when working with Iceberg in my opinion (though not in all situations).

If we automatically register the functions for users when the catalog is instantiated, does that have any performance impact or downsides?

Also, for naming, is it possible that we might clash with some other name (if we registered the functions for people)?

I’m working on function registration in Spark so we should ideally get our names somewhat stabilized.

I would propose `iceberg_bucket` but need to better understand how it is used.

Reply:
Flink supports 4 properties categories of functions:
1. Temporary system functions
2. System functions
3. Temporary catalog functions
4. Catalog functions

If user is using Ambiguous Function Reference, The resolution order is:

1. Temporary system function
2. System function
3. Temporary catalog function, in the current catalog and current database of the session
4. Catalog function, in the current catalog and current database of the session

For details https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/dev/table/functions/overview/


In the code, these two functions are registered in the catalog, so 
1. If the user uses iceberg's catalog, they can use these two functions directly. 
2. If the user does not use iceberg's catalog, they need to register these two functions before using them (eg written in test).

Reply:
eg:
```java
  // first, create iceberg catalog
  StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
  tEnv = StreamTableEnvironment.create(env);
  tEnv.executeSql("" +
          "CREATE CATALOG hive_catalog WITH (\n" +
          "    'type'='iceberg',\n" +
          "    'catalog-type'='hive',\n" +
          .....
          "  )");
    
    // second, call function by Precise Function Reference
    tEnv.executeSql("select hive_catalog.db.buckets(200, 12341)");        // you can use function directly by Precise Function Reference under hive_catalog

    // third, or call function by Ambiguous Function Reference
    tEnv.executeSql("use catalog hive_catalog");
    tEnv.executeSql("select buckets(200, 12341)");        // you can use function directly by Ambiguous Function Reference under hive_catalog
```

Reply:
For the ambiguous function reference here, for clarification, this will work?

Once this is merged, this might be something we want to document, or at least point user's to the Flink documentation for function resolution when we mention these UDFs.

Reply:
It will work. Because if `use catalog hive_catalog` is not used, then the current catalog is the `default_catalog`, and will can't find the `buckets` function.

Reply:
Can you separate the tests for registering these as UDFs from the tests for calling them through the catalog? I want to make sure we have separate tests for those two cases.